fix(server): handle `=` in cookie values correctly

aeneasr opened the pull request:
This patch resolves an issue where using `=` in cookie values would cause the cookie value to contain only the partial value.

Closes https://github.com/microsoft/playwright/issues/11612

<!--
Thank you for your Pull Request. Please link to the issue this PR addresses.
-->

<!-- PR Checklist
- The issue has been triaged and positive signals were received from maintainers
- Existing tests pass
- New tests are added
- Relevant documentation is changed or added
-->

aeneasr commented:
Not sure why the tests fails, they do not seem related to this change?

yury-s commented on packages/playwright-core/src/server/fetch.ts:
```diff
@@ -517,7 +517,11 @@ function toHeadersArray(rawHeaders: string[]): types.HeadersArray {
-  const pairs = header.split(';').filter(s => s.trim().length > 0).map(p => p.split('=').map(s => s.trim()));
+  const pairs = header.split(';').filter(s => s.trim().length > 0).map(p => {
+    const value = p.slice(p.indexOf('=') + 1).trim();
+    const key = p.slice(0, p.indexOf('=')).trim();
```
This will not work for the attributes that have only name (no value and no '='), e.g. `HttpOnly`. Please fix this and extend your test case to cover this.

aeneasr replied:
Thank you for the fast review! I will!

Do you have a tip for me where I can execute this test only? Running the full test suite locally causes a lot of test failures. Thank you :)

aeneasr replied:
Nevermind, this does the trick:

```
npm run ctest -- tests/global-fetch-cookie.spec.ts
```

aeneasr replied:
Would it be acceptable to export `parseCookie` and test it individually? It seems a bit difficult to test it well given how it is integrated

yury-s replied:
We normally don't do unit tests. You can read parsed cookie via `request.storageState()`, similar to this test: https://github.com/microsoft/playwright/blob/6a7a2971f2e40442d720f42412d299c504fcbd71/tests/global-fetch-cookie.spec.ts#L175-L188

aeneasr replied:
Thank you, that was very helpful!

yury-s commented on tests/global-fetch-cookie.spec.ts:
```diff
@@ -67,6 +67,20 @@ it('should store cookie from Set-Cookie header', async ({ request, server }) =>
 });
 
+it('should store cookie from Set-Cookie header with trailing equal sign', async ({ request, server }) => {
+  // https://github.com/microsoft/playwright/issues/11612
```
let's convert it to the following annotation(we started using this form recently) :

```js
 it.info().annotations.push({ type: 'issue', description: 'https://github.com/microsoft/playwright/issues/11612' });
```

yury-s reviewed (changes requested):
Thanks for the comment, the fix looks good modulo the problem with name-only attributes.

yury-s commented on packages/playwright-core/src/server/fetch.ts:
```diff
@@ -517,7 +517,10 @@ function toHeadersArray(rawHeaders: string[]): types.HeadersArray {
-  const pairs = header.split(';').filter(s => s.trim().length > 0).map(p => p.split('=').map(s => s.trim()));
+  const pairs = header.split(';').filter(s => s.trim().length > 0).map(p => {
+    const [key, ...rest] = p.split('=');
+    return [key.trim(), rest.map(s => s.trim()).join('=')];
```
This will break on `a=b == c`, I also liked the previous version of this code more as it didn't do unnecessary split/join, perhaps you could resurrect that and just add a check if '=' is present?

aeneasr replied:
Sorry, I just pushed a change which fixes the trim issue. I also added a test for it! Is that solution still acceptable? It's basically like the one we had before, it just handles the rest of the values a bit differently.

yury-s replied:
I liked the version without extra `.split` calls more as the intent of the code was more clear, I'd rather you just modified it to something like the following
```js
  let key;
  let value;
  const separatorPos = p.indexOf('=');
  if (separatorPos === -1) {
    key = p.trim();
  } else {
    key = p.slice(0, separatorPos).trim();
    value = p.slice(separatorPos + 1).trim();
  }
  return [key, value];
```
but it is up to you.

aeneasr replied:
Ah indeed, it's definitely more readable! I adjusted your version and added a few comments to make clear what this is for

aeneasr committed:
fix(server): handle `=` in cookie values correctly

This patch resolves an issue where using `=` in cookie values would cause the cookie value to contain only the partial value.

Closes https://github.com/microsoft/playwright/issues/11612

yury-s reviewed (approved):
lgtm, but looks like now the lint is unhappy (not sure why it didn't complain before your change), initializing key/value with empty string should help I believe.

aeneasr commented:
Thanks, resolved :)

aeneasr commented:
Thank you @yury-s for help, rewview & merge :) I looked around a bit but couldn't find some type of nightly build that might include this change. Is there anything available to using nightly / builds from main? If not it would be awesome if this fix could be in the next patch release :)

aeneasr commented:
Nevermind, it appears there are nightlies on NPM :)
